File: src/models/black_scholes.cpp
#include <cmath>
#include <stdexcept>
#include <thales/models/black_scholes.hpp>

namespace thales {
namespace models {

double BlackScholes::call_price(double S, double K, double r, double sigma,
                                double T) {
    if (T <= 0) {
        // For expired options, return intrinsic value
        return std::max(0.0, S - K);
    }

    double d1Val = d1(S, K, r, sigma, T);
    double d2Val = d2(S, K, r, sigma, T);

    return S * normal_cdf(d1Val) - K * std::exp(-r * T) * normal_cdf(d2Val);
}

double BlackScholes::put_price(double S, double K, double r, double sigma,
                               double T) {
    if (T <= 0) {
        // For expired options, return intrinsic value
        return std::max(0.0, K - S);
    }

    double d1Val = d1(S, K, r, sigma, T);
    double d2Val = d2(S, K, r, sigma, T);

    return K * std::exp(-r * T) * normal_cdf(-d2Val) - S * normal_cdf(-d1Val);
}

double BlackScholes::call_delta(double S, double K, double r, double sigma,
                                double T) {
    if (T <= 0) {
        // For expired options, delta is either 0 or 1
        return (S > K) ? 1.0 : 0.0;
    }

    return normal_cdf(d1(S, K, r, sigma, T));
}

double BlackScholes::put_delta(double S, double K, double r, double sigma,
                               double T) {
    if (T <= 0) {
        // For expired options, delta is either -1 or 0
        return (S < K) ? -1.0 : 0.0;
    }

    return normal_cdf(d1(S, K, r, sigma, T)) - 1.0;
}

double BlackScholes::gamma(double S, double K, double r, double sigma,
                           double T) {
    if (T <= 0) {
        // For expired options, gamma is 0
        return 0.0;
    }

    double d1Val = d1(S, K, r, sigma, T);

    return normal_pdf(d1Val) / (S * sigma * std::sqrt(T));
}

double BlackScholes::vega(double S, double K, double r, double sigma,
                          double T) {
    if (T <= 0) {
        // For expired options, vega is 0
        return 0.0;
    }

    double d1Val = d1(S, K, r, sigma, T);

    return S * std::sqrt(T) * normal_pdf(d1Val) /
           100.0;  // Divided by 100 to get the change per 1% change in
                   // volatility
}

double BlackScholes::call_theta(double S, double K, double r, double sigma,
                                double T) {
    if (T <= 0) {
        // For expired options, theta is 0
        return 0.0;
    }

    double d1Val = d1(S, K, r, sigma, T);
    double d2Val = d2(S, K, r, sigma, T);

    double term1 = -S * sigma * normal_pdf(d1Val) / (2.0 * std::sqrt(T));
    double term2 = -r * K * std::exp(-r * T) * normal_cdf(d2Val);

    return (term1 + term2) / 365.0;  // Divided by 365 to get the daily theta
}

double BlackScholes::put_theta(double S, double K, double r, double sigma,
                               double T) {
    if (T <= 0) {
        // For expired options, theta is 0
        return 0.0;
    }

    double d1Val = d1(S, K, r, sigma, T);
    double d2Val = d2(S, K, r, sigma, T);

    double term1 = -S * sigma * normal_pdf(d1Val) / (2.0 * std::sqrt(T));
    double term2 = r * K * std::exp(-r * T) * normal_cdf(-d2Val);

    return (term1 + term2) / 365.0;  // Divided by 365 to get the daily theta
}

double BlackScholes::call_rho(double S, double K, double r, double sigma,
                              double T) {
    if (T <= 0) {
        // For expired options, rho is 0
        return 0.0;
    }

    double d2Val = d2(S, K, r, sigma, T);

    return K * T * std::exp(-r * T) * normal_cdf(d2Val) /
           100.0;  // Divided by 100 to get the change per 1% change in interest
                   // rate
}

double BlackScholes::put_rho(double S, double K, double r, double sigma,
                             double T) {
    if (T <= 0) {
        // For expired options, rho is 0
        return 0.0;
    }

    double d2Val = d2(S, K, r, sigma, T);

    return -K * T * std::exp(-r * T) * normal_cdf(-d2Val) /
           100.0;  // Divided by 100 to get the change per 1% change in interest
                   // rate
}

double BlackScholes::call_implied_volatility(double price, double S, double K,
                                             double r, double T, double epsilon,
                                             int max_iterations) {
    if (T <= 0) {
        throw std::invalid_argument("Time to expiration must be positive");
    }

    // Check for arbitrage bounds
    double intrinsicValue = std::max(0.0, S - K * std::exp(-r * T));
    if (price < intrinsicValue) {
        throw std::invalid_argument("Option price violates arbitrage bounds");
    }

    // Initial guess for implied volatility
    double sigma = 0.2;  // Start with 20% volatility

    // Newton-Raphson method
    for (int i = 0; i < max_iterations; ++i) {
        double priceEstimate = call_price(S, K, r, sigma, T);
        double diff = price - priceEstimate;

        if (std::abs(diff) < epsilon) {
            return sigma;
        }

        double vegaEstimate = vega(S, K, r, sigma, T);
        if (vegaEstimate == 0.0) {
            throw std::runtime_error(
                "Vega is zero, cannot calculate implied volatility");
        }

        sigma +=
            diff / (vegaEstimate *
                    100.0);  // Multiply by 100 because vega is per 1% change

        // Ensure sigma stays within reasonable bounds
        if (sigma <= 0.001) {
            sigma = 0.001;
        } else if (sigma > 10.0) {
            throw std::runtime_error(
                "Implied volatility calculation did not converge");
        }
    }

    throw std::runtime_error(
        "Implied volatility calculation did not converge within maximum "
        "iterations");
}

double BlackScholes::put_implied_volatility(double price, double S, double K,
                                            double r, double T, double epsilon,
                                            int max_iterations) {
    if (T <= 0) {
        throw std::invalid_argument("Time to expiration must be positive");
    }

    // Check for arbitrage bounds
    double intrinsicValue = std::max(0.0, K * std::exp(-r * T) - S);
    if (price < intrinsicValue) {
        throw std::invalid_argument("Option price violates arbitrage bounds");
    }

    // Initial guess for implied volatility
    double sigma = 0.2;  // Start with 20% volatility

    // Newton-Raphson method
    for (int i = 0; i < max_iterations; ++i) {
        double priceEstimate = put_price(S, K, r, sigma, T);
        double diff = price - priceEstimate;

        if (std::abs(diff) < epsilon) {
            return sigma;
        }

        double vegaEstimate = vega(S, K, r, sigma, T);
        if (vegaEstimate == 0.0) {
            throw std::runtime_error(
                "Vega is zero, cannot calculate implied volatility");
        }

        sigma +=
            diff / (vegaEstimate *
                    100.0);  // Multiply by 100 because vega is per 1% change

        // Ensure sigma stays within reasonable bounds
        if (sigma <= 0.001) {
            sigma = 0.001;
        } else if (sigma > 10.0) {
            throw std::runtime_error(
                "Implied volatility calculation did not converge");
        }
    }

    throw std::runtime_error(
        "Implied volatility calculation did not converge within maximum "
        "iterations");
}

double BlackScholes::d1(double S, double K, double r, double sigma, double T) {
    return (std::log(S / K) + (r + sigma * sigma / 2.0) * T) /
           (sigma * std::sqrt(T));
}

double BlackScholes::d2(double S, double K, double r, double sigma, double T) {
    return d1(S, K, r, sigma, T) - sigma * std::sqrt(T);
}

double BlackScholes::normal_cdf(double x) {
    // Approximation of the cumulative distribution function of the standard
    // normal distribution This is the Abramowitz and Stegun approximation

    const double b1 = 0.31938153;
    const double b2 = -0.356563782;
    const double b3 = 1.781477937;
    const double b4 = -1.821255978;
    const double b5 = 1.330274429;
    const double p = 0.2316419;
    const double c = 0.39894228;

    if (x >= 0.0) {
        double t = 1.0 / (1.0 + p * x);
        return 1.0 - c * std::exp(-x * x / 2.0) * t *
                         (t * (t * (t * (t * b5 + b4) + b3) + b2) + b1);
    } else {
        double t = 1.0 / (1.0 - p * x);
        return c * std::exp(-x * x / 2.0) * t *
               (t * (t * (t * (t * b5 + b4) + b3) + b2) + b1);
    }
}

double BlackScholes::normal_pdf(double x) {
    // Probability density function of the standard normal distribution
    return (1.0 / std::sqrt(2.0 * M_PI)) * std::exp(-0.5 * x * x);
}

}  // namespace models
}  // namespace thales

File: python/src/thales_backtester/models/black_scholes.py
import numpy as np
from scipy.stats import norm
import math
from typing import Dict, Literal, Optional, Tuple, Union


class BlackScholesModel:
    """
    Python implementation of the Black-Scholes-Merton option pricing model.
    This serves as a fallback when the C++ bindings are not available.
    """
    
    @staticmethod
    def d1(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the d1 parameter for Black-Scholes formula.
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            d1 parameter value
        """
        if T <= 0 or sigma <= 0:
            # Handle edge cases to avoid math domain errors
            return float('nan')
            
        return (np.log(S / K) + (r + 0.5 * sigma**2) * T) / (sigma * np.sqrt(T))
    
    @staticmethod
    def d2(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the d2 parameter for Black-Scholes formula.
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            d2 parameter value
        """
        if T <= 0 or sigma <= 0:
            # Handle edge cases to avoid math domain errors
            return float('nan')
            
        return BlackScholesModel.d1(S, K, r, sigma, T) - sigma * np.sqrt(T)
    
    @staticmethod
    def call_price(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the theoretical price of a European call option.
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Theoretical call option price
        """
        if T <= 0:  # Handle expired options
            return max(0, S - K)
            
        if sigma <= 0:  # Handle zero volatility
            return max(0, S - K * np.exp(-r * T))
            
        d1_val = BlackScholesModel.d1(S, K, r, sigma, T)
        d2_val = BlackScholesModel.d2(S, K, r, sigma, T)
        
        if np.isnan(d1_val) or np.isnan(d2_val):
            return float('nan')
            
        return S * norm.cdf(d1_val) - K * np.exp(-r * T) * norm.cdf(d2_val)
    
    @staticmethod
    def put_price(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the theoretical price of a European put option.
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Theoretical put option price
        """
        if T <= 0:  # Handle expired options
            return max(0, K - S)
            
        if sigma <= 0:  # Handle zero volatility
            return max(0, K * np.exp(-r * T) - S)
            
        d1_val = BlackScholesModel.d1(S, K, r, sigma, T)
        d2_val = BlackScholesModel.d2(S, K, r, sigma, T)
        
        if np.isnan(d1_val) or np.isnan(d2_val):
            return float('nan')
            
        return K * np.exp(-r * T) * norm.cdf(-d2_val) - S * norm.cdf(-d1_val)
    
    @staticmethod
    def call_delta(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the delta of a call option.
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Call option delta
        """
        if T <= 0:  # Handle expired options
            return 1.0 if S > K else 0.0
            
        d1_val = BlackScholesModel.d1(S, K, r, sigma, T)
        
        if np.isnan(d1_val):
            return float('nan')
            
        return norm.cdf(d1_val)
    
    @staticmethod
    def put_delta(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the delta of a put option.
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Put option delta
        """
        if T <= 0:  # Handle expired options
            return -1.0 if S < K else 0.0
            
        d1_val = BlackScholesModel.d1(S, K, r, sigma, T)
        
        if np.isnan(d1_val):
            return float('nan')
            
        return norm.cdf(d1_val) - 1.0
    
    @staticmethod
    def gamma(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the gamma of an option (same for calls and puts).
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Option gamma
        """
        if T <= 0 or sigma <= 0:  # Handle edge cases
            return 0.0
            
        d1_val = BlackScholesModel.d1(S, K, r, sigma, T)
        
        if np.isnan(d1_val):
            return float('nan')
            
        return norm.pdf(d1_val) / (S * sigma * np.sqrt(T))
    
    @staticmethod
    def vega(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the vega of an option.
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Option vega (per 1% change in volatility)
        """
        if T <= 0 or sigma <= 0:  # Handle edge cases
            return 0.0
            
        d1_val = BlackScholesModel.d1(S, K, r, sigma, T)
        
        if np.isnan(d1_val):
            return float('nan')
            
        # Scale by 0.01 to get the change for a 1% change in volatility
        return S * np.sqrt(T) * norm.pdf(d1_val) * 0.01
    
    @staticmethod
    def call_theta(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the theta of a call option (time decay).
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Call option theta (per year)
        """
        if T <= 0 or sigma <= 0:  # Handle edge cases
            return 0.0
            
        d1_val = BlackScholesModel.d1(S, K, r, sigma, T)
        d2_val = BlackScholesModel.d2(S, K, r, sigma, T)
        
        if np.isnan(d1_val) or np.isnan(d2_val):
            return float('nan')
            
        common_term = -(S * sigma * norm.pdf(d1_val)) / (2 * np.sqrt(T))
        specific_term = -r * K * np.exp(-r * T) * norm.cdf(d2_val)
        
        # Return daily theta (divided by 365)
        return (common_term + specific_term) / 365.0
    
    @staticmethod
    def put_theta(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the theta of a put option (time decay).
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Put option theta (per year)
        """
        if T <= 0 or sigma <= 0:  # Handle edge cases
            return 0.0
            
        d1_val = BlackScholesModel.d1(S, K, r, sigma, T)
        d2_val = BlackScholesModel.d2(S, K, r, sigma, T)
        
        if np.isnan(d1_val) or np.isnan(d2_val):
            return float('nan')
            
        common_term = -(S * sigma * norm.pdf(d1_val)) / (2 * np.sqrt(T))
        specific_term = r * K * np.exp(-r * T) * norm.cdf(-d2_val)
        
        # Return daily theta (divided by 365)
        return (common_term + specific_term) / 365.0
    
    @staticmethod
    def call_rho(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the rho of a call option.
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Call option rho (per 1% change in interest rate)
        """
        if T <= 0:  # Handle expired options
            return 0.0
            
        d2_val = BlackScholesModel.d2(S, K, r, sigma, T)
        
        if np.isnan(d2_val):
            return float('nan')
            
        # Scale by 0.01 to get the change for a 1% change in interest rate
        return K * T * np.exp(-r * T) * norm.cdf(d2_val) * 0.01
    
    @staticmethod
    def put_rho(S: float, K: float, r: float, sigma: float, T: float) -> float:
        """
        Calculate the rho of a put option.
        
        Args:
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            sigma: Volatility of the underlying (annualized)
            T: Time to expiration in years
            
        Returns:
            Put option rho (per 1% change in interest rate)
        """
        if T <= 0:  # Handle expired options
            return 0.0
            
        d2_val = BlackScholesModel.d2(S, K, r, sigma, T)
        
        if np.isnan(d2_val):
            return float('nan')
            
        # Scale by 0.01 to get the change for a 1% change in interest rate
        return -K * T * np.exp(-r * T) * norm.cdf(-d2_val) * 0.01
    
    @staticmethod
    def call_implied_volatility(price: float, S: float, K: float, r: float, T: float, 
                               epsilon: float = 0.0001, max_iterations: int = 100) -> float:
        """
        Calculate the implied volatility of a call option using the Newton-Raphson method.
        
        Args:
            price: Market price of the option
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            T: Time to expiration in years
            epsilon: Desired precision for implied volatility
            max_iterations: Maximum number of iterations
            
        Returns:
            Implied volatility
        """
        if T <= 0:  # Handle expired options
            return float('nan')
            
        if price <= 0:  # Handle invalid prices
            return float('nan')
            
        # Check for arbitrage bounds
        intrinsic_value = max(0, S - K * np.exp(-r * T))
        if price < intrinsic_value:
            return float('nan')  # Price violates arbitrage bounds
            
        # Initial guess for implied volatility
        sigma = 0.2  # Start with 20% volatility
        
        for i in range(max_iterations):
            price_diff = BlackScholesModel.call_price(S, K, r, sigma, T) - price
            
            if abs(price_diff) < epsilon:
                return sigma
                
            vega_val = BlackScholesModel.vega(S, K, r, sigma, T) * 100  # Adjust vega scale
            
            if abs(vega_val) < 1e-10:
                return float('nan')
                
            sigma = sigma - price_diff / vega_val
            
            # Ensure sigma stays within reasonable bounds
            if sigma <= 0.001:
                sigma = 0.001
            elif sigma > 10:  # 1000% volatility cap
                return float('nan')
                
        # Failed to converge
        return float('nan')
    
    @staticmethod
    def put_implied_volatility(price: float, S: float, K: float, r: float, T: float,
                              epsilon: float = 0.0001, max_iterations: int = 100) -> float:
        """
        Calculate the implied volatility of a put option using the Newton-Raphson method.
        
        Args:
            price: Market price of the option
            S: Current stock price
            K: Option strike price
            r: Risk-free interest rate (annualized)
            T: Time to expiration in years
            epsilon: Desired precision for implied volatility
            max_iterations: Maximum number of iterations
            
        Returns:
            Implied volatility
        """
        if T <= 0:  # Handle expired options
            return float('nan')
            
        if price <= 0:  # Handle invalid prices
            return float('nan')
            
        # Check for arbitrage bounds
        intrinsic_value = max(0, K * np.exp(-r * T) - S)
        if price < intrinsic_value:
            return float('nan')  # Price violates arbitrage bounds
            
        # Initial guess for implied volatility
        sigma = 0.2  # Start with 20% volatility
        
        for i in range(max_iterations):
            price_diff = BlackScholesModel.put_price(S, K, r, sigma, T) - price
            
            if abs(price_diff) < epsilon:
                return sigma
                
            vega_val = BlackScholesModel.vega(S, K, r, sigma, T) * 100  # Adjust vega scale
            
            if abs(vega_val) < 1e-10:
                return float('nan')
                
            sigma = sigma - price_diff / vega_val
            
            # Ensure sigma stays within reasonable bounds
            if sigma <= 0.001:
                sigma = 0.001
            elif sigma > 10:  # 1000% volatility cap
                return float('nan')
                
        # Failed to converge
        return float('nan')
